Sources/Workspace/WorkspaceConfiguration.swift

```diff
@@ -50,39 +47,279 @@ extension Workspace {
-            sharedCacheDirectory: AbsolutePath? = .none
+            resolvedVersionsFile: AbsolutePath,
+            sharedCacheDirectory: AbsolutePath?,
+            sharedConfigurationDirectory: AbsolutePath?
         ) {
             self.workingDirectory = workingDirectory
             self.editsDirectory = editsDirectory
-            self.resolvedVersionsFilePath = resolvedVersionsFilePath
+            self.resolvedVersionsFile = resolvedVersionsFile
             self.sharedCacheDirectory = sharedCacheDirectory
+            self.sharedConfigurationDirectory = sharedConfigurationDirectory
         }
 
         /// Create a new workspace location.
         ///
         /// - Parameters:
         ///   - rootPath: Path to the root of the package, from which other locations can be derived.
-        public init(forRootPackage rootPath: AbsolutePath) {
+        public init(forRootPackage rootPath: AbsolutePath, fileSystem: FileSystem) {
             self.init(
-                workingDirectory: rootPath.appending(component: ".build"),
-                editsDirectory: rootPath.appending(component: "Packages"),
-                resolvedVersionsFilePath: rootPath.appending(component: "Package.resolved")
+                workingDirectory: DefaultLocations.workingDirectory(forRootPackage: rootPath),
+                editsDirectory: DefaultLocations.editsDirectory(forRootPackage: rootPath),
+                resolvedVersionsFile: DefaultLocations.resolvedVersionsFile(forRootPackage: rootPath),
+                sharedCacheDirectory: fileSystem.swiftPMCacheDirectory,
+                sharedConfigurationDirectory: fileSystem.swiftPMConfigDirectory
             )
         }
     }
+}
+
+// MARK: - Default locations
+extension Workspace {
+    /// Workspace default locations utilities
+    public struct DefaultLocations {
+        public static func workingDirectory(forRootPackage rootPath: AbsolutePath) -> AbsolutePath {
+            rootPath.appending(component: ".build")
+        }
+
+        public static func editsDirectory(forRootPackage rootPath: AbsolutePath) -> AbsolutePath {
+            rootPath.appending(component: "Packages")
+        }
+
+        public static func resolvedVersionsFile(forRootPackage rootPath: AbsolutePath) -> AbsolutePath {
+            rootPath.appending(component: "Package.resolved")
+        }
+
+        public static func configurationDirectory(forRootPackage rootPath: AbsolutePath) -> AbsolutePath {
+            rootPath.appending(components: ".swiftpm", "configuration")
+        }
+
+        public static func mirrorsConfigurationFile(forRootPackage rootPath: AbsolutePath) -> AbsolutePath {
+            mirrorsConfigurationFile(at: configurationDirectory(forRootPackage: rootPath))
+        }
+
+        public static func mirrorsConfigurationFile(at path: AbsolutePath) -> AbsolutePath {
+            path.appending(component: "mirrors.json")
+        }
+    }
+}
+
+// MARK: - Mirrors
+
+// FIXME
```

Comment:
still need to figure final name here

Reply:
How about defining the `Configuration` namespace as a nested enumeration in `Workspace`?

Reply:
right, the issue is we already have `Workspace.Configuration` which is the current mirrors configuration which this deprecates https://github.com/apple/swift-package-manager/pull/3670/files#diff-8ea56cbcd3f938d28b68dee635b50b8afb4616cbc002044be94cf268743b7323R327

Reply:
Ah, didn't see that it was deprecated and not removed. Maybe this is too janky, but could we deprecate just the members and keep `Workspace.Configuration` around just for the namespace?

Aside from that, another possibility might be `Workspace.Config`. Short of using completely different terminology, at least that has the benefit of brevity.

Reply:
okay I went for the janky option, we can deprecate the class and turn it into an enum later